### ISSUE TITLE
kube-proxy: log warning and/or Event if a NodePort overlaps with on-node ephemeral range

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#111823]
https://github.com/kubernetes/kubernetes/pull/111697 documents not to do this.  Would be good to make it visible when/if it does happen.

Something like this on each sync:
* open and parse /proc/sys/net/ipv4/ip_local_port_range
* compare all NodePorts and HCNP against this
* if overlap, log and Event

@danwinship 

Also need it in KPNG, I guess.

**Comment by aojea** [comment]
> * if overlap, log and Event

this should be rate limited ... several escalations recently because of Events filling etcd :upside_down_face:

**Comment by aojea** [comment]
also, NodePorts affects to all the cluster nodes, we don't want one event per node, in a large cluster this can be risky ... I'll start with the log only

**Comment by thockin** [comment]
Nobody looks at logs :(

OK to start there, but maybe we should use a lease and only whichever
node holds the lease is allowed to send events?  This presumes that
all nodes are configured the same, which might not be true but close
enough?

**Comment by aojea** [comment]
> Nobody looks at logs :(

that is sad by true ... 

> This presumes that all nodes are configured the same, which might not be true but close
enough?
 
heh, another fact ... 

most probably the people that check the logs are the ones that have the nodes differently configured 😄

**Comment by danwinship** [comment]
> This presumes that all nodes are configured the same, which might not be true but close enough?

You could use a rule like "only check NodePort/ephemeral overlap for services where `epslice.Endpoints[0]` is on this node" so that every node would check _some_ services

**Comment by thockin** [comment]
We'd still storm Events for daemonsets and large-N deployments.

**Comment by haoruan** [comment]
In kube-proxy, the eventRecorder has some [deduplication logic](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/383-new-event-api-ga-graduation/README.md#changes-in-eventrecorder) to reduce the event emission, so there will be only one event up to 30mins, but still one event per node, is it good to go?

**Comment by danwinship** [comment]
> We'd still storm Events for daemonsets and large-N deployments.

No, I mean, for every service, there would be exactly 1 node that was allowed to emit the event for that service. eg, if the EndpointSlice has

    endpoints:
      - addresses: ["10.0.2.1"]
        nodeName: "node-2"
      - addresses: ["10.0.5.3"]
        nodeName: "node-5"
      - addresses: ["10.0.10.2"]
        nodeName: "node-10"

then node-2 would do the check and emit the event if needed, and every other node would just create the nodeport without complaint.

**Comment by thockin** [comment]
`nodeName` is optional and for manual services doesn't make sense.
Why not just use a lease?

**Comment by danwinship** [comment]
It's optional because some endpoints aren't on nodes. Kube-proxy requires that `nodeName` be set for pod-based endpoints. (Things like local traffic policy won't work right if it's not set.)

> Why not just use a lease?

The idea was in response to your comment about "presumes that all nodes are configured the same, which might not be true"

**Comment by thockin** [comment]
I get what you are saying but I am not sure it's practically much better.
If the ephemeral range differs across nodes, it is non-deterministic
whether the service which has a colliding node-port happens to land on a
node that has the "bad" range.  And now you need to decide who handles
node-port collisions for corner cases.

E.g.

Service a has nodeport 32222
Service b has nodeport 33333
Service c has a nodeport 44444

The 0th endpoint for "a" lands on node n1 whose ephemeral ports are 32k-60k
The 0th endpoint for "b" lands on a node n2 whose ephemeral ports are
40k-60k
There are no endpoints for "c" with nodeName set.

In your model (assuming I didn't mess it up), this will result in 0 events,
even though b and c are both problematic on at least SOME nodes.   Since we
don't know the differences between the nodes a priori, I think the best we
can do is periodically rotate which node will report events and
rate-limit.  "n1" would have reported b and c.  "n2" would have reported c.

**Comment by aojea** [comment]
one important detail .. the NodePort range is a flag in the apiserver and its value is not exposed, so kube-proxy can't really know what is the NodePort range unless we pass it via a per-node flag , that makes the odds of having different configurations still higher :upside_down_face:

**Comment by danwinship** [comment]
> If the ephemeral range differs across nodes, it is non-deterministic whether the service which has a colliding node-port happens to land on a node that has the "bad" range.

Right, the idea was that _eventually_ you'd get a bad node/port combo though, and it would report it, and eventually the admin would fix things and then they'd stay fixed. (Once we start warning about this, it should eventually become something that's only a problem for new clusters, not for long-term clusters.)

> I think the best we can do is periodically rotate which node will report events

Ah, I was thinking more like "node A gets the lease and holds onto it for as long as the node remains up". If the lease rotates then that has the same "eventually" property I was going for.

> kube-proxy can't really know what is the NodePort range unless we pass it via a per-node flag

If kube-proxy knew the whole range then it could just error out at startup, yes, but it can still warn about individual services without knowing that.

> that makes the odds of having different configurations still higher

I don't think so. It's plausible that a cluster would have "infra" nodes with a larger ephemeral port range than ordinary "worker" nodes, but if different nodes of the same type have different sysctl configurations then you probably have larger problems with your cluster deployment strategy...

**Comment by khenidak** [comment]
shouldn't that be part of something like node-problem-detector?

**Comment by shaneutt** [comment]
This is getting stale, but it still seems relevant. I'm going to re-triage it, possibly even work on it myself I'm not certain yet.

/unassign @haoruan 
/assign @shaneutt

**Comment by shaneutt** [comment]
This is valid, and is something we could work on.

/triage accepted
/help

However the priority is not particularly high.

/priority backlog
